File: cli/src/tools/shell.ts
/**
 * Shell command execution tools
 */

import { exec } from 'child_process';
import { promisify } from 'util';
import { ToolResult, ok, err } from './index.js';

const execAsync = promisify(exec);

/**
 * Run a shell command
 */
export async function runCommand(cmd: string, timeout = 30000): Promise<ToolResult> {
  try {
    const isWindows = process.platform === 'win32';
    const shell = isWindows ? 'cmd.exe' : '/bin/sh';
    const shellArg = isWindows ? '/C' : '-c';

    const { stdout, stderr } = await execAsync(cmd, {
      shell,
      timeout,
      maxBuffer: 10 * 1024 * 1024, // 10MB
    });

    if (stderr && stderr.trim()) {
      return ok(`${stdout}\n[stderr]: ${stderr}`);
    }
    return ok(stdout || '(no output)');
  } catch (error: any) {
    if (error.killed) {
      return err(`Command timed out after ${timeout}ms`);
    }
    const stderr = error.stderr || '';
    const stdout = error.stdout || '';
    return {
      success: false,
      output: stdout,
      error: `Exit code: ${error.code || -1}\n${stderr}`,
    };
  }
}

/**
 * Get environment variable
 */
export function getEnv(varName: string): ToolResult {
  const value = process.env[varName];
  if (value !== undefined) {
    return ok(value);
  }
  return err(`Environment variable not set: ${varName}`);
}

/**
 * Set environment variable (for current process)
 */
export function setEnv(varName: string, value: string): ToolResult {
  process.env[varName] = value;
  return ok(`Set ${varName}=${value}`);
}

/**
 * Get current working directory
 */
export function getCwd(): ToolResult {
  return ok(process.cwd());
}

/**
 * Change current working directory
 */
export function setCwd(path: string): ToolResult {
  try {
    process.chdir(path);
    return ok(`Changed directory to: ${path}`);
  } catch (error: any) {
    return err(`Failed to change directory: ${error.message}`);
  }
}

/**
 * Shell tool definitions for agents
 */
export const shellTools = [
  {
    name: 'run_command',
    description: 'Execute shell command',
    parameters: {
      type: 'object',
      properties: {
        cmd: { type: 'string', description: 'Command to execute' },
        timeout: { type: 'number', description: 'Timeout in ms (default 30000)' },
      },
      required: ['cmd'],
    },
  },
  {
    name: 'get_env',
    description: 'Get environment variable',
    parameters: {
      type: 'object',
      properties: {
        var: { type: 'string', description: 'Variable name' },
      },
      required: ['var'],
    },
  },
  {
    name: 'set_env',
    description: 'Set environment variable',
    parameters: {
      type: 'object',
      properties: {
        var: { type: 'string', description: 'Variable name' },
        value: { type: 'string', description: 'Value to set' },
      },
      required: ['var', 'value'],
    },
  },
  {
    name: 'get_cwd',
    description: 'Get current working directory',
    parameters: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'set_cwd',
    description: 'Change working directory',
    parameters: {
      type: 'object',
      properties: {
        path: { type: 'string', description: 'Directory path' },
      },
      required: ['path'],
    },
  },
];

File: src/tools/shell.rs
//! Shell command execution tools

use super::ToolResult;
use std::process::Command;

/// Run a shell command
pub fn run_command(cmd: &str) -> ToolResult {
    let output = if cfg!(windows) {
        Command::new("cmd")
            .args(["/C", cmd])
            .output()
    } else {
        Command::new("sh")
            .args(["-c", cmd])
            .output()
    };

    match output {
        Ok(output) => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            let stderr = String::from_utf8_lossy(&output.stderr);

            if output.status.success() {
                if stderr.is_empty() {
                    ToolResult::ok(stdout.to_string())
                } else {
                    ToolResult::ok(format!("{}\n[stderr]: {}", stdout, stderr))
                }
            } else {
                ToolResult {
                    success: false,
                    output: stdout.to_string(),
                    error: Some(format!(
                        "Exit code: {}\n{}",
                        output.status.code().unwrap_or(-1),
                        stderr
                    )),
                }
            }
        }
        Err(e) => ToolResult::err(format!("Failed to execute command: {}", e)),
    }
}

/// Get environment variable
pub fn get_env(var: &str) -> ToolResult {
    match std::env::var(var) {
        Ok(value) => ToolResult::ok(value),
        Err(_) => ToolResult::err(format!("Environment variable not set: {}", var)),
    }
}

/// Set environment variable (for current process)
pub fn set_env(var: &str, value: &str) -> ToolResult {
    std::env::set_var(var, value);
    ToolResult::ok(format!("Set {}={}", var, value))
}

/// Get current working directory
pub fn get_cwd() -> ToolResult {
    match std::env::current_dir() {
        Ok(path) => ToolResult::ok(path.to_string_lossy().to_string()),
        Err(e) => ToolResult::err(format!("Failed to get working directory: {}", e)),
    }
}

/// Change current working directory
pub fn set_cwd(path: &str) -> ToolResult {
    match std::env::set_current_dir(path) {
        Ok(()) => ToolResult::ok(format!("Changed directory to: {}", path)),
        Err(e) => ToolResult::err(format!("Failed to change directory: {}", e)),
    }
}
